astropy/utils/tests/test_data.py

```diff
@@ -430,62 +904,456 @@ def __init__(self, data):
+        check_free_space_in_dir(tmpdir, 1_000_000_000_000_000_000)
+
+
+def test_get_free_space_file_directory(tmpdir):
+    fn = tmpdir / "file"
+    with open(fn, "w"):
+        pass
+    with pytest.raises(OSError):
+        get_free_space_in_dir(fn)
+    assert get_free_space_in_dir(tmpdir) > 0
+
+
+def test_download_file_bogus_settings(invalid_urls, temp_cache):
+    u = next(invalid_urls)
+    with pytest.raises(ValueError):
+        download_file(u, sources=[])
+
+
+def test_download_file_local_directory(tmpdir):
+    with pytest.raises(urllib.request.URLError):
+        download_file(url_to(tmpdir))
+
+
+def test_download_file_schedules_deletion(valid_urls):
+    u, c = next(valid_urls)
+    f = download_file(u)
+    assert f in _tempfilestodel
+    # how to test deletion actually occurs?
+
+
+def test_clear_download_cache_refuses_to_delete_outside_the_cache(tmpdir):
+    fn = os.path.abspath(os.path.join(tmpdir, "file"))
+    with open(fn, "w") as f:
+        f.write("content")
+    assert os.path.exists(fn)
+    with pytest.raises(RuntimeError):
+        clear_download_cache(fn)
+    assert os.path.exists(fn)
+
+
+def test_check_download_cache_finds_unreferenced_files(temp_cache, valid_urls):
+    u, c = next(valid_urls)
+    download_file(u, cache=True)
+    with _cache(write=True) as (dldir, urlmap):
+        del urlmap[u]
+    with pytest.raises(ValueError):
+        check_download_cache()
+    clear_download_cache()
+
+
+def test_check_download_cache_finds_missing_files(temp_cache, valid_urls):
+    u, c = next(valid_urls)
+    os.remove(download_file(u, cache=True))
+    with pytest.raises(ValueError):
+        check_download_cache()
+    clear_download_cache()
+
+
+def test_check_download_cache_finds_bogus_entries(temp_cache, valid_urls):
+    u, c = next(valid_urls)
+    download_file(u, cache=True)
+    with _cache(write=True) as (dldir, urlmap):
+        bd = os.path.join(dldir, "bogus")
+        os.mkdir(bd)
+        bf = os.path.join(bd, "file")
+        with open(bf, "wt") as f:
+            f.write("bogus file that exists")
+        urlmap[u] = bf
+    with pytest.raises(ValueError):
+        check_download_cache()
+    clear_download_cache()
+
+
+def test_check_download_cache_finds_bogus_hashes(temp_cache, valid_urls):
+    u, c = next(valid_urls)
+    fn = download_file(u, cache=True)
+    with open(fn, "w") as f:
+        f.write("bogus contents")
+    with pytest.raises(ValueError):
+        check_download_cache(check_hashes=True)
+    clear_download_cache()
+
+
+def test_download_cache_update_doesnt_damage_cache(temp_cache, valid_urls):
+    u, _ = next(valid_urls)
+    download_file(u, cache=True)
+    download_file(u, cache="update")
+
+
+def test_cache_dir_is_actually_a_file(tmpdir, valid_urls):
+    def check_quietly_ignores_bogus_cache():
+        with catch_warnings(CacheMissingWarning) as w:
+            assert not get_cached_urls()
+        assert any(wi.category == CacheMissingWarning for wi in w)
```

Comment:
Why not use `pytest.warns`?

Reply:
Because I didn't know about it? Probably the astropy helper was written before it existed.

Reply:
Yes, there is even #6761 about this but `catch_warnings` would be around for a while longer, if not forever.